stage #3/java/module #2. REST API Basics/task02/service/src/main/java/by/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,204 @@
+                    .update(giftCertificateMapper.giftCertificateDtoToGiftCertificate(giftCertificateDto));
+            return setTagsForDto(giftCertificateMapper.giftCertificateToGiftCertificateDto(updatedCertificate));
+        } else {
+            throw new ServiceException(ErrorCode.NOT_FIND_CERTIFICATE_BY_ID,
+                    ErrorCode.NOT_FIND_CERTIFICATE_BY_ID.getMessage(),
+                    Set.of(new ErrorMessage(
+                            "id",
+                            String.valueOf(giftCertificateDto.getId()),
+                            ErrorCode.NOT_FIND_CERTIFICATE_BY_ID.getMessage()
+                    )));
+        }
+    }
+
+    @Override
+    public void delete(Integer id) {
+        if (!(giftCertificateDao.deleteById(id))) {
+            throw new ServiceException(ErrorCode.NOT_FIND_CERTIFICATE_BY_ID,
+                    ErrorCode.NOT_FIND_CERTIFICATE_BY_ID.getMessage(),
+                    Set.of(new ErrorMessage("id", String.valueOf(id), ErrorCode.NOT_FIND_CERTIFICATE_BY_ID.getMessage()))
+            );
+        }
+    }
+
+
+    @Override
+    public List<GiftCertificateDto> findAll(DtoGiftCertificateRequestParam CertificateRequestDto,
+                                            PaginationDto paginationDto) {
+        baseValidator.dtoValidator(paginationDto);
+        baseValidator.dtoValidator(CertificateRequestDto);
+        GiftCertificateRequestParam giftCertificateRequestParam = giftCertificateRequestMapper.dtoGiftCertificateRequestToGiftCertificate(CertificateRequestDto);
+        Pagination pagination = paginationMapper.paginationDtoToPagination(paginationDto);
+        List<GiftCertificateDto> dtoList = giftCertificateDao.findAll(giftCertificateRequestParam, pagination)
+                .stream()
+                .map(giftCertificateMapper::giftCertificateToGiftCertificateDto)
+                .collect(Collectors.toList());
+        dtoList.forEach(this::setTagsForDto);
+        return dtoList;
+    }
+
+    @Override
+    public GiftCertificateDto partUpdate(GiftCertificateDto giftCertificateDto) {
+        baseValidator.dtoValidatorForPartUpdate(giftCertificateDto);
+        if (giftCertificateDto.getName() != null) {
+            checkIsGiftCertificateNameExists(giftCertificateDto);
+        }
+        checkAndUpdateTags(giftCertificateDto.getTags());
+        GiftCertificate certificateForUpdate = giftCertificateMapper.giftCertificateDtoToGiftCertificate(giftCertificateDto);
+        GiftCertificateDto updatedCertificate = giftCertificateMapper.
+                giftCertificateToGiftCertificateDto(giftCertificateDao.partUpdate(certificateForUpdate));
+        return setTagsForDto(updatedCertificate);
+    }
+
+    //if user send tag with id
+    private void checkAndUpdateTags(List<TagDto> listDto) {
+        if (listDto != null) {
+            listDto.stream()
+                    .filter(tagDto -> tagDto.getName() != null)
```

Comment:
Sooo, do you think that there cannot be tags with just 'id' or just 'name'? Also there is a requirement:
If tag doesn't exist, it must be created